### PR TITLE
Abstract Practice

### PR DESCRIPTION
**Pull request by sperBruno**

### CODE REVIEW

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.fundacionjala.abstrac;
+
+/**
+ * Created by Bruno on 6/16/2016.
+ */
+public class Circle extends Operation {
+    private double ratio;
+    private final double pi = 3.14;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: please use the ```java Math.PI ``` constant variable

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    private final int DIAMETER = 2;
+
+    public Circle(double ratio) {
+        this.ratio = ratio;
+    }
+
+    public double calculatePerimeter() {
+        perimeter = DIAMETER * ratio;
+        return perimeter;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: You are using an unnecessary variable
```java
return  DIAMETER * ratio;
```

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    public double calculateArea() {
+        area = DIAMETER * pi * ratio;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: You are using an unnecessary variable
```java
return DIAMETER * pi * ratio;
```

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
+package com.fundacionjala.abstrac;
+
+/**
+ * Created by Bruno on 6/16/2016.
+ */
+public class Circle extends Operation {
+    private double ratio;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: This attribute should be final

File: src/main/java/com/fundacionjala/abstrac/Rectangle.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package com.fundacionjala.abstrac;
+
+/**
+ * Created by Bruno on 6/16/2016.
+ */
+public class Rectangle extends Operation {
+    private int base, higth;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Rectangle.java]: These attributes should be final

File: src/main/java/com/fundacionjala/abstrac/Operation.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package com.fundacionjala.abstrac;
+
+/**
+ * Created by Bruno on 6/16/2016.
+ */
+abstract class Operation {
+    protected double area;
+
+    protected double perimeter;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Operation.java]: I think the area and perimeter variables are not necessary

File: src/main/java/com/fundacionjala/abstrac/Rectangle.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+    public Rectangle(int base, int hight) {
+        this.base = base;
+        this.higth = hight;
+    }
+
+    public double calculatePerimeter() {
+        perimeter = ((2 * base) + (2 * higth));
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Rectangle.java]: Magic Number

File: src/test/java/com/fudacionjala/abstrac/TestCircle.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
+package com.fudacionjala.abstrac;
+
+import com.fundacionjala.abstrac.Circle;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestCircle.java]: Organize your imports

File: src/test/java/com/fudacionjala/abstrac/TestRectangle.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package com.fudacionjala.abstrac;
+
+
+import com.fundacionjala.abstrac.Rectangle;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestRectangle.java]: Organize your imports

File: src/test/java/com/fudacionjala/abstrac/TestSquare.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
+package com.fudacionjala.abstrac;
+
+import com.fundacionjala.abstrac.Square;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestSquare.java]: Organize your imports

File: src/main/java/com/fundacionjala/abstrac/Rectangle.java (hunk 7 above)
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Rectangle.java]: Optimize the formula
```java
return FACTOR_FORMULA * (base + higth);
```

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+        this.ratio = ratio;
+    }
+
+    /**
+     * This method calculate the circle perimeter.
+     *
+     * @return the perimeter.
+     */
+    public double calculatePerimeter() {
+        return Constants.OPERATION_VARIABLE * PI * ratio;
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: Please add a static import

File: src/main/java/com/fundacionjala/abstrac/Constants.java [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+ * This Constants class has variables that are going to be used through different classes.
+ *
+ * @author Bruno Barrios Vargas
+ */
+public class Constants {
+
+    public static final int DELTA = 0;
+
+    public static final int OPERATION_VARIABLE = 2;
+
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Constants.java]: Don't forget to add a private constructor for Constants class (static variables and methods)

File: src/test/java/com/fudacionjala/abstrac/TestCircle.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        circle = new Circle(3.0);
+    }
+
+    /**
+     * This method verifies the correctness answer of Circle perimeter operation.
+     */
+    @Test
+    public void testCalculateCirclePerimeter() {
+        double expectedPerimeterCircle = 18.84955592153876;
+        assertEquals(expectedPerimeterCircle, circle.calculatePerimeter(), Constants.DELTA);
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestCircle.java]: Please add a static import

File: src/test/java/com/fudacionjala/abstrac/TestSquare.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+        double expectedResult = 8.0;
+        assertEquals(expectedResult, square.calculatePerimeter());
+    }
+
+    /**
+     * This method verifies the correctness answer of Square area operation.
+     */
+    @Test
+    public void testCalculatePerimeter() {
+        double expectedResult = 4.0;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestSquare.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestSquare.java [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    public void setUp() {
+        square = new Square(2);
+    }
+
+    /**
+     * This method verifies the correctness answer of Square perimeter operation.
+     */
+    @Test
+    public void testCalculateSquareArea() {
+        double expectedResult = 8.0;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestSquare.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestRectangle.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+        double expectedResult = 12.0;
+        assertEquals(expectedResult, rectangle.calculatePerimeter());
+    }
+
+    /**
+     * This method verifies the correctness answer of Rectangle area operation.
+     */
+    @Test
+    public void testCalculateAreaRectangle() {
+        double expectedResult = 8.0;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestRectangle.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestRectangle.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+    public void testInstanceRectangle() {
+        rectangle = new Rectangle(2, 4);
+    }
+
+    /**
+     * This method verifies the correctness answer of Rectangle perimeter operation.
+     */
+    @Test
+    public void testCalculatePerimeterRectangle() {
+        double expectedResult = 12.0;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestRectangle.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestRectangle.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public class TestRectangle {
+    private Rectangle rectangle;
+
+    /**
+     * This method instances the requeriment of each junit methods.
+     */
+    @Before
+    public void testInstanceRectangle() {
+        rectangle = new Rectangle(2, 4);
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestRectangle.java]: Magic number

File: src/test/java/com/fudacionjala/abstrac/TestSquare.java [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+public class TestSquare {
+
+    private Square square;
+
+    /**
+     * This method instances the requeriment of each junit methods.
+     */
+    @Before
+    public void setUp() {
+        square = new Square(2);
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestSquare.java]: Magic number

File: src/test/java/com/fudacionjala/abstrac/TestCircle.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     * This method verifies the correctness answer of Circle area operation.
+     */
+    @Test
+    public void testCalculateCircleArea() {
+        double expectedResult = 28.274333882308138;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestCircle.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestCircle.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    public void setUp() {
+        circle = new Circle(3.0);
+    }
+
+    /**
+     * This method verifies the correctness answer of Circle perimeter operation.
+     */
+    @Test
+    public void testCalculateCirclePerimeter() {
+        double expectedPerimeterCircle = 18.84955592153876;
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestCircle.java]: This variable should be final

File: src/test/java/com/fudacionjala/abstrac/TestCircle.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+public class TestCircle {
+
+    private Circle circle;
+
+    /**
+     * This method instances the requeriment of each junit methods.
+     */
+    @Before
+    public void setUp() {
+        circle = new Circle(3.0);
```
Comment (carledriss) [inline src/test/java/com/fudacionjala/abstrac/TestCircle.java]: magic number

File: src/main/java/com/fundacionjala/abstrac/Circle.java [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    /**
+     * This method calculate the circle area.
+     *
+     * @return the perimeter.
+     */
+    public double calculateArea() {
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Circle.java]: this should be
```java
/**
 * {@inheritDoc}
 */
```

**Comment by carledriss** [comment]
After minor changes
Looks good to me.

**Comment by sperBruno** [comment]
@carledriss please review the last commit

File: src/main/java/com/fundacionjala/abstrac/Operation.java [hunk 24]
```diff
@@ -0,0 +1,14 @@
+package com.fundacionjala.abstrac;
+
+/**
+ * Created by Bruno on 6/16/2016.
+ */
+abstract class Operation {
```
Comment (carledriss) [inline src/main/java/com/fundacionjala/abstrac/Operation.java]: Is this class used?
Remove it.

**Comment by carledriss** [comment]
Looks good to me!

**Pull request merged**